### ISSUE TITLE
Inconsistency in expectations between ProgressBar#advance and Meter#sample

### ISSUE THREAD

**Issue by doudou** [piotrmurach/tty-progressbar#15]
I started to use tty-progressbar to do a byte-related progress, but noticed really weird rates (not matching at all the overall mean rates I would calculate)

So far, I am assuming that when using the _rate tokens, one should get a rate calculation while passing actual units (not units per second) to ProgressBar's #advance and #current= methods.

From reading the code and the specs, it seems that Meter#sample expects to be given a rate (NOT an absolute advance). I say that since it's only averaging the sampled values instead of taking into account the sampling times.

But ProgressBar#advance is passing an absolute progress (in unit, not in unit/s).

I would propose a PR fixing this, but I'm not sure whether the "right" fix is to change Meter or ProgressBar (IMO, Meter could be changed to do more interesting things with rate calculations). If you're OK with it, I would change ProgressBar#sample to take the time and the absolute value (instead of progress) and update Meter to take into account the sample's time.

**Comment by piotrmurach** [comment]
Hi Sylvain, good to hear that you are using tty libraries.

It's been a while since I wrote this logic. The `Meter#sample` takes the current time and the relative progress, i.e. the current value for `.advance` call. The `.advance` is always thought as relative step. Thus the sample expects the current time and the relative advance value. Then depending whether the sampling is within the current 1 second interval(which is configurable), it either gets added to current sample or new sample is created by increasing the marker. In other words,

```ruby
progress = TTY::Progress.new ....
progress.start
progress.advance(5)
```

will setup internal clocks for both the progress and the meter. Therefore, each time 

```ruby
# in first second
progress.advance(10) 
progress.advance(10)

# in second second
progress.advance(10) 
```
Then I would expect `rate`  would be `5` because it only shows the current second rate, that  is, total advances in given second over the total number of samples. Wherease the `mean_rate` I would expect to be `15` as a total average of all current samples. Therefore it maybe the case that the formatter for bytes needs to be changed to use `mean_rate`? In any case, I think the behaviour should probably be changed in the Meter. I will happily review any PRs.

**Comment by doudou** [comment]
I'm not sure I understand ....

I'm going to try to give an example of what behaviour I would expect

```
progress = TTY::Progress.new ....
progress.start                     # t0 = 0
progress.advance(5)          # t0+0.1     mean_rate=50   rate=50   (50/0.1) between [t0+0;t0+0.1]
progress.advance(5)          # t0+0.5     mean_rate=20   rate=20   (10/0.5) between [t0+0;t0+0.5]
progress.advance(5)          # t0+1        mean_rate=15   rate=15   (15/1) between [t0+0;t0+1]
progress.advance(5)          # t0+1.5     mean_rate=13 (20/1.5)  rate=10   (10/1) between [t0+0.5;t0+1.5]
progress.advance(10)          # t0+2      mean_rate=15 (30/2)     rate=15   (15/1) between [t0+1;t0+2]
```

**Comment by piotrmurach** [comment]
Makes sense, sorry for confusing matters in the first place. My initial implementation probably was meant to track rate in this fashion but got convoluted in the process. Ok, let's update the Meter as per your initial suggestion and take into account sampling times. Would you have time to work on this?

**Comment by doudou** [comment]
> Would you have time to work on this?

Yes.

**Comment by piotrmurach** [comment]
Cool, I'm gonna be flying in two days but will be available over the weekend to respond/review or merge anything.

**Comment by piotrmurach** [comment]
Released `v0.10.0`, thanks for fixing this.
